gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java

```diff
@@ -875,6 +878,59 @@ public boolean delete(Path hadoopPath, boolean recursive) throws IOException {
     return status.toArray(new FileStatus[0]);
   }
 
+  class GCSFilesListIterator implements RemoteIterator<LocatedFileStatus> {
+    private String nextPageToken;
+    private boolean recursive;
+    private Path hadoopPath;
+    private Iterator<FileInfo> currentPageIterator;
+
+    public GCSFilesListIterator(Path hadoopPath, boolean recursive) throws IOException {
+      this.hadoopPath = hadoopPath;
+      this.nextPageToken = null;
+      this.recursive = recursive;
+      //load first page with token null.
+      loadNextBatch();
+    }
+
+    @Override
+    public boolean hasNext() throws IOException {
+      if (currentPageIterator == null) return false;
+      if (!currentPageIterator.hasNext() && nextPageToken != null) {
+        loadNextBatch();
+      }
+      return currentPageIterator.hasNext();
+    }
+
+    /**
+     * Load next batch with pageToken, in case of token null. it will always return first page
+     *
+     * @throws IOException
+     */
+    private void loadNextBatch() throws IOException {
+      ListPage<FileInfo> currentBatch = getGcsFs().listFileInfoForPrefixPage(getGcsPath(hadoopPath), this.nextPageToken, recursive);
+      if (currentBatch != null && currentBatch.getItems() != null && currentBatch.getItems().size() != 0) {
+        this.nextPageToken = currentBatch.getNextPageToken();
+        this.currentPageIterator = currentBatch.getItems().iterator();
+      }
+    }
+
+    @Override
+    public LocatedFileStatus next() throws IOException {
+      checkArgument(hasNext(), "No next found");
```

Comment:
`checkArgument` throws `IllegalArgumentException`, in current case we want `IllegalStateException` 

Can we replace with 
```
Preconditions.checkState(this.hasNext(), "No more items in iterator");
```